File: packages/distance/src/edit/jaro.ts
/**
 * Jaro and Jaro-Winkler similarity algorithms.
 *
 * Jaro measures similarity between two strings by considering matching characters
 * and transpositions. Jaro-Winkler extends Jaro with a prefix bonus.
 *
 * Time: O(m * n)
 */

/**
 * Compute Jaro similarity between two strings.
 *
 * J(S1, S2) = (1/3) * (m/|S1| + m/|S2| + (m - t/2) / m)
 *
 * where m = number of matching characters (within window),
 *       t = number of transpositions among matching characters.
 *
 * @param a - First string
 * @param b - Second string
 * @returns Jaro similarity in [0, 1]
 */
export function jaro(a: string, b: string): number {
  const aLen = a.length;
  const bLen = b.length;

  if (aLen === 0 && bLen === 0) return 1;
  if (aLen === 0 || bLen === 0) return 0;

  const matchDistance = Math.floor(Math.max(aLen, bLen) / 2) - 1;
  if (matchDistance < 0) return 0;

  const aMatches = new Uint8Array(aLen);
  const bMatches = new Uint8Array(bLen);

  let matches = 0;
  let transpositions = 0;

  for (let i = 0; i < aLen; i++) {
    const start = Math.max(0, i - matchDistance);
    const end = Math.min(i + matchDistance + 1, bLen);

    for (let j = start; j < end; j++) {
      if (bMatches[j] || a.charCodeAt(i) !== b.charCodeAt(j)) continue;
      aMatches[i] = 1;
      bMatches[j] = 1;
      matches++;
      break;
    }
  }

  if (matches === 0) return 0;

  let k = 0;
  for (let i = 0; i < aLen; i++) {
    if (!aMatches[i]) continue;
    while (!bMatches[k]) k++;
    if (a.charCodeAt(i) !== b.charCodeAt(k)) transpositions++;
    k++;
  }

  return (matches / aLen + matches / bLen + (matches - transpositions / 2) / matches) / 3;
}

/**
 * Options for Jaro-Winkler similarity.
 */
export interface IJaroWinklerOptions {
  /**
   * Weight applied to the common prefix bonus.
   * @default 0.1
   */
  prefixWeight?: number;

  /**
   * Maximum length of common prefix to consider.
   * @default 4
   */
  maxPrefix?: number;
}

/**
 * Compute Jaro-Winkler similarity between two strings.
 *
 * JW(S1, S2) = Jaro(S1, S2) + l * p * (1 - Jaro(S1, S2))
 *
 * where l = length of common prefix (up to maxPrefix),
 *       p = prefixWeight.
 *
 * @param a - First string
 * @param b - Second string
 * @param options - Configuration
 * @returns Jaro-Winkler similarity in [0, 1]
 */
export function jaroWinkler(a: string, b: string, options: IJaroWinklerOptions = {}): number {
  const p = options.prefixWeight ?? 0.1;
  const maxPrefix = options.maxPrefix ?? 4;

  const jaroScore = jaro(a, b);

  let l = 0;
  const minLen = Math.min(a.length, b.length, maxPrefix);
  while (l < minLen && a.charCodeAt(l) === b.charCodeAt(l)) l++;

  return jaroScore + l * p * (1 - jaroScore);
}

File: packages/distance-wasm/src/edit/jaro.rs
/// Jaro and Jaro-Winkler similarity algorithms.
///
/// Time: O(m * n)

pub fn jaro(a: &str, b: &str) -> f64 {
    let a_bytes = a.as_bytes();
    let b_bytes = b.as_bytes();
    let a_len = a_bytes.len();
    let b_len = b_bytes.len();

    if a_len == 0 && b_len == 0 { return 1.0; }
    if a_len == 0 || b_len == 0 { return 0.0; }

    let max_len = a_len.max(b_len);
    let match_distance = if max_len <= 1 {
        if a_len == b_len && a_bytes == b_bytes { return 1.0; } else { return 0.0; }
    } else {
        max_len / 2 - 1
    };

    let mut a_matches = vec![false; a_len];
    let mut b_matches = vec![false; b_len];
    let mut matches = 0u32;
    let mut transpositions = 0u32;

    // Find matching characters
    for i in 0..a_len {
        let start = i.saturating_sub(match_distance);
        let end = (i + match_distance + 1).min(b_len);
        for j in start..end {
            if b_matches[j] || a_bytes[i] != b_bytes[j] { continue; }
            a_matches[i] = true;
            b_matches[j] = true;
            matches += 1;
            break;
        }
    }

    if matches == 0 { return 0.0; }

    // Count transpositions
    let mut k = 0usize;
    for i in 0..a_len {
        if !a_matches[i] { continue; }
        while !b_matches[k] { k += 1; }
        if a_bytes[i] != b_bytes[k] { transpositions += 1; }
        k += 1;
    }

    let m = matches as f64;
    (m / a_len as f64 + m / b_len as f64 + (m - transpositions as f64 / 2.0) / m) / 3.0
}

pub fn jaro_winkler(a: &str, b: &str) -> f64 {
    let p = 0.1;
    let max_prefix = 4;

    let jaro_score = jaro(a, b);

    let a_bytes = a.as_bytes();
    let b_bytes = b.as_bytes();
    let min_len = a_bytes.len().min(b_bytes.len()).min(max_prefix);
    let mut l = 0usize;
    while l < min_len && a_bytes[l] == b_bytes[l] {
        l += 1;
    }

    jaro_score + l as f64 * p * (1.0 - jaro_score)
}
